### PR TITLE
Simplify crop_or_pad with deepcopy

### PR DESCRIPTION
**Pull request by yasahi-hpc**
The PR aims at cleaning up `crop_or_pad` function in the same way as #167.
This allows `crop_or_pad` can assume that the data is already allocated

- [x] views are passed by `const&` in `crop_or_pad`
- [x] Extend `create_layout` to handle `std::array<std::size_t, N>`
- [x] Updating unit_tests and internal usage based on this change

### CODE REVIEW

File: common/src/KokkosFFT_padding.hpp [hunk 1]
```diff
@@ -349,7 +235,11 @@ void crop_or_pad(const ExecutionSpace& exec_space, const InViewType& in,
       "type (float/double), the same layout (LayoutLeft/LayoutRight), and the "
       "same rank. ExecutionSpace must be accessible to the data in InViewType "
       "and OutViewType.");
-  crop_or_pad_impl(exec_space, in, out, s);
+  if constexpr (InViewType::rank() == 1) {
+    crop_or_pad_1D(exec_space, in, out);
+  } else {
+    crop_or_pad_ND(exec_space, in, out);
+  }
```
Comment (tpadioleau) [inline common/src/KokkosFFT_padding.hpp:242]:
```suggestion
  crop_or_pad(exec_space, in, out);
```

Can we implement such a function ? Either with a new function or merge the two existing if possible ?

File: common/src/KokkosFFT_utils.hpp [hunk 2]
```diff
@@ -209,13 +209,35 @@ auto extract_extents(const ViewType& view) {
...
+  using extents_type         = std::array<IntType, rank>;
+
+  extents_type extents;
+  for (std::size_t i = 0; i < rank; i++) {
+    extents.at(i) = std::min(in.extent(i), out.extent(i));
+  }
+  return extents;
+}
+
+template <typename Layout, typename ContainerType>
```
Comment (tpadioleau) [inline common/src/KokkosFFT_utils.hpp:230]: Is this change necessary for the current PR ? Does `create_layout` need to be such generic ?

File: common/src/KokkosFFT_padding.hpp [hunk 3]
```diff
@@ -107,240 +105,128 @@ void crop_or_pad_impl(const ExecutionSpace& exec_space, const InViewType& in,
...
-      });
-}
-
-template <typename ExecutionSpace, typename InViewType, typename OutViewType>
-void crop_or_pad_impl(const ExecutionSpace& exec_space, const InViewType& in,
-                      OutViewType& out, shape_type<5> s) {
-  constexpr std::size_t DIM = 5;
+void crop_or_pad_ND(const ExecutionSpace& exec_space, const InViewType& in,
+                    const OutViewType& out) {
+  [[maybe_unused]] auto upper =
```
Comment (tpadioleau) [inline common/src/KokkosFFT_padding.hpp:110]: Does the compiler still complain without `maybe_unused` ? To me it seems to be used in all branches.

File: common/src/KokkosFFT_padding.hpp [hunk 4]
```diff
@@ -107,240 +105,128 @@ void crop_or_pad_impl(const ExecutionSpace& exec_space, const InViewType& in,
...
-  );
-
-  Kokkos::parallel_for(
-      "KokkosFFT::crop_or_pad", range,
-      KOKKOS_LAMBDA(int i0, int i1, int i2, int i3, int i4, int i5) {
-        for (int i6 = 0; i6 < n6; i6++) {
-          for (int i7 = 0; i7 < n7; i7++) {
-            out(i0, i1, i2, i3, i4, i5, i6, i7) =
-                in(i0, i1, i2, i3, i4, i5, i6, i7);
+  if constexpr (InViewType::rank() == 2) {
```
Comment (tpadioleau) [inline common/src/KokkosFFT_padding.hpp:120]: Two related questions about this `crop_or_pad` function but not directly linked to this PR:

- Could we avoid manual unpacking of integers by using the conversion functions `Kokkos::Array` <-> `std::array` ?
- Why don't we use `Kokkos::deep_copy` with `Kokkos::subview` ?

**Review by tpadioleau — CHANGES_REQUESTED** [review]
I think it is a good thing, it simplifies the reasoning about the functions `crop_or_pad`.

File: fft/src/KokkosFFT_Plans.hpp [hunk 5]
```diff
@@ -314,10 +314,13 @@ class Plan {
...
         typename KokkosFFT::Impl::manageable_view_type<OutViewType>::type;
-    ManagableInViewType in_s;
+
     InViewType in_tmp;
     if (m_is_crop_or_pad_needed) {
-      KokkosFFT::Impl::crop_or_pad(m_exec_space, in, in_s, m_shape);
+      using LayoutType = typename ManagableInViewType::array_layout;
+      ManagableInViewType const in_s(
+          "in_s", KokkosFFT::Impl::create_layout<LayoutType>(m_shape));
+      KokkosFFT::Impl::crop_or_pad(m_exec_space, in, in_s);
```
[the remark this answers is not in the document]
Comment (yasahi-hpc) [reply fft/src/KokkosFFT_Plans.hpp:323]: That is true. 
This part should be improved.

For the moment, I would rather prefer to make `in_tmp` managed.

File: common/unit_test/Test_Utils.cpp [hunk 6]
```diff
@@ -526,6 +526,59 @@ TEST(ExtractExtents, 1Dto8D) {
   EXPECT_EQ(KokkosFFT::Impl::extract_extents(view8D), ref_extents8D);
 }
 
+TEST(ShrankExtents, 1Dto8D) {
```
[the remark this answers is not in the document]
Comment (yasahi-hpc) [reply common/unit_test/Test_Utils.cpp:529]: May be not that important, just for completeness

File: common/src/KokkosFFT_padding.hpp (hunk 4 above)
Comment (yasahi-hpc) [reply common/src/KokkosFFT_padding.hpp:120]: That is a good point. I will give it a try

File: common/src/KokkosFFT_padding.hpp (hunk 3 above)
Comment (yasahi-hpc) [reply common/src/KokkosFFT_padding.hpp:110]: Maybe OK now. I will try to remove

**Comment by yasahi-hpc** [comment]
@tpadioleau 
Thank you for your review. 
I have fixed accordingly.

I do not like the current implementation of `shrank_slicers`. 
Do you have any improvements in mind?

**Comment by tpadioleau** [comment]
> @tpadioleau Thank you for your review. I have fixed accordingly.
> 
> I do not like the current implementation of `shrank_slicers`. Do you have any improvements in mind?

what do you think of [this](https://godbolt.org/#z:OYLghAFBqd5QCxAYwPYBMCmBRdBLAF1QCcAaPECAMzwBtMA7AQwFtMQByARg9KtQYEAysib0QXACx8BBAKoBnTAAUAHpwAMvAFYTStJg1DIApACYAQuYukl9ZATwDKjdAGFUtAK4sGe1wAyeAyYAHI%2BAEaYxCAAzABspAAOqAqETgwe3r56KWmOAkEh4SxRMYl2mA4ZQgRMxARZPn5ctpj2BQy19QRFYZHRcRV1DU05rQojvcH9pYPxAJS2qF7EyOwc5rHByN5YANQmsW71xEwAnkfYJhoAgls7e5iHxwTnSZgA%2BgRnhApXN3uZm2DF2XgORzcXkctEIl1i1zugKY0NQ%2BwA1qh0ZiFJ8FF4IgA3PCYADuEBRRHM8Wp%2B2CpH2lNQ1NpKwICyOVjuhNQeHQGKxOM%2BWEwSU%2BaCS5wpqJZZni%2B0wqiqDKZsvl9MZMrlrOhHNiXPudwImBYSQMxpebjeH2YbH22CVyGhNSSTHWDOtjFYzwAkgwAGok0kAFXemA9YdtzwA8tDA2TQx9SID9qm0%2BnJugQCA0gAvL4EfahBmZ7N5gsAOir%2Bx9/wRgJ5fP2IqSEvOnwEeIJxLJCggaAYk3tjudAiErvWtMVVTxE/DKfTi6Xy7TA6HfvjIbDtI1a8LsYIm8TmB1BGTdxXl8ve/2pZApwukLv5e%2BDNCALl%2B1QhOiBiSSWCYBPmITBgAyc9bivKDFzvYIsFUPFMAARy8RhJ2OWsqwrK4FkOAB2A10yZW8CU%2BYIXgAEQFbFUi7IkgwgBclw1O8WCYdEvldPBiAgDQSwILMQGATACEhWsrggb9fyYf9AOA0CMgWBYsL1Qi02I/EIg7aFKOooVNJ7ckmMXNl%2BMEtiOM%2BLieL428BOzYTRIwutsEkn9iD/ACjHksCBCUlTOWMzEaNxFtxVQSUIGnZBZzdcMSK00yErIhhVMBEw8Io9KjRNM0mAtSFPSjYcqlHLo5wjG1vRrAMg2PSqvTtA8jzDCCr2fPB82%2BfYKJ9ABZXSuABbleX5ZBiAijtiCsph0H7AQhwdUrOnHOKp0dWL3X2G8Nzq7dP3pYzoO2hb9zjPaPlPXCMrU1M90VJJiDswSX0LM4GHRXTdoTMNs3e9EIDSi80y8NIjAVVRjUEXFPV0u8H3hNwOq6s99n%2BgF9WyyC00VKGCBhsMIbxutbv2fgnogZGCzpXSNE5GnITRwx0XpvBrGsa6CKOnHIcYfGK3yiA8Gu2IqNY4IhYYCtcb5oWli/aFpd5wQ5aB7HUwyrKkWB1MWzbDsGDowy%2B2izb4o1JKZehtrjqg1j2K%2BODFUQlC0JPY50frLnMrVzXAQ4JZaE4ABWXg/G4XhUE4Nx2csW8VjWZ4th4UgCE0AOlnREB4jMCsAA5YliDQNDzyRJDwjRJFiYPpCDjhJDD9PSCjjheAUEA%2BLTjgtCWOBYBgRAUFQU06GichKDQEf6BiYApDMPg6GNYh24gCIm4iYJ6nOTgU435hiHOaMIm0Kou5Tye2EEaMGFobfu94LAIi8YATloWh24j0gsDYoxxHvr/uKnzwD%2BD%2BWhSDRWhBsMBwRjR1zAbCCIZwD4eCwE3H4eAWA714O5CIqRMAURNIYYAsIjDpyWFQAwwAFCbmjDaLBMhBAiDEOwKQDD5BKDUE3XQrQDCkNMJYaw%2Bg8ARHbrAKMIBgo4lIO5TgZdi4Z2bkkToH8AC0mYjgUX4VYSwZgwFSWIMQPkmBRGAzaB0DILgGDuE8M0fwVi%2BglDKLkVI6QBBjBaMkFxnQHEDBiBMdoQCxxTHcXoSo1Qgk9B8XMPxthgk2PGLEyJMxHGDC4EsBQCd1gSEDiHRu/8W77FUHneIKj4iSH2MAZAyB9hSArGYfYEBcCEBIIcYEaTeBdx7pnEAwc%2BJ1wbqQTBOcKx4WBMHMwXBYi6MkAATmDlwGZpBw56M4G3DuqcyGkD7oPNkSRoTj3mlPaIoRvScCKSUspFSqk1MkHU3gmB8BEEMVmVo/BGGiHEKwt57CVDqH/tw0gpIzhJCwTkjgocllNxbgePZhZUBUEKcU0p5TKnVNqfUiAHgjlPWTgsDpZClgIEwLNQYpj%2Bm8EwbEPCdS8JlLwsUvCczK551aMsyOqzbDrM6QorOVKKxcHiPEGZMyuDB3GXMspiQ66xDySs1uGz75KX0JwMwsr2Xyu5Uq9yaRnCSCAA)

**Comment by yasahi-hpc** [comment]
Cool!
I will try option 2

**Comment by yasahi-hpc** [comment]
With this, I have errors

```C++
template <typename ExecutionSpace, typename InViewType, typename OutViewType,
          std::size_t... Is>
void crop_or_pad_impl(const ExecutionSpace& exec_space, const InViewType& in,
                      const OutViewType& out, std::index_sequence<Is...>) {
  constexpr std::size_t rank = InViewType::rank();
  using extents_type = std::array<std::size_t, rank>;

  extents_type extents;
  for (std::size_t i = 0; i < rank; i++) {
    extents.at(i) = std::min(in.extent(i), out.extent(i));
  }

  auto sub_in  = Kokkos::subview(in, std::make_pair(0, std::get<Is>(extents))...);
  auto sub_out = Kokkos::subview(out, std::make_pair(0, std::get<Is>(extents))...);
  Kokkos::deep_copy(exec_space, sub_out, sub_in);
}

template <typename ExecutionSpace, typename InViewType, typename OutViewType>
void crop_or_pad(const ExecutionSpace& exec_space, const InViewType& in,
                 const OutViewType& out) {
  static_assert(
      KokkosFFT::Impl::are_operatable_views_v<ExecutionSpace, InViewType,
                                              OutViewType>,
      "crop_or_pad: InViewType and OutViewType must have the same base "
      "floating point "
      "type (float/double), the same layout (LayoutLeft/LayoutRight), and the "
      "same rank. ExecutionSpace must be accessible to the data in InViewType "
      "and OutViewType.");
  crop_or_pad_impl(exec_space, in, out, std::make_index_sequence<InViewType::rank()>{});
}
```

Errors are something like

```bash
/work/04/jh230064a/i18048/tests/develop/kokkos-fft/tpls/kokkos/core/src/impl/Kokkos_ViewMapping.hpp(3352): error: no instance of constructor "Kokkos::Impl::ViewOffset<Dimension, Kokkos::LayoutRight, std::enable_if_t<<expression>, void>>::ViewOffset [with Dimension=Kokkos::Impl::ViewDimension<0UL, 0UL, 0UL, 0UL, 0UL>]" matches the argument list
            argument types are: (const Kokkos::Impl::ViewOffset<Kokkos::Impl::ViewDimension<0UL, 0UL, 0UL, 0UL, 0UL>, Kokkos::LayoutRight, void>, const Kokkos::Impl::SubviewExtents<5U, 0U>)
      dst.m_impl_offset = dst_offset_type(src.m_impl_offset, extents);
                          ^
          detected during:
            instantiation of "void Kokkos::Impl::ViewMapping<std::enable_if_t<<expression>, void>, SrcTraits, Args...>::assign(Kokkos::Impl::ViewMapping<DstTraits, void> &, const Kokkos::Impl::ViewMapping<SrcTraits, void> &, Args...) [with SrcTraits=Kokkos::ViewTraits<Kokkos::complex<float> *****, Kokkos::LayoutRight, execution_space>, Args=<std::pair<int, std::size_t>, std::pair<int, std::size_t>, std::pair<int, std::size_t>, std::pair<int, std::size_t>, std::pair<int, std::size_t>>, DstTraits=Kokkos::ViewTraits<Kokkos::complex<float> *****, Kokkos::LayoutRight, Kokkos::Device<Kokkos::CudaSpace::execution_space, Kokkos::CudaSpace::memory_space>, Kokkos::MemoryTraits<0U>>]" at line 1372 of /work/04/jh230064a/i18048/tests/develop/kokkos-fft/tpls/kokkos/core/src/Kokkos_View.hpp
            instantiation of "Kokkos::View<DataType, Properties...>::View(const Kokkos::View<RT, RP...> &, Arg0, Args...) [with DataType=Kokkos::complex<float> *****, Properties=<Kokkos::LayoutRight, Kokkos::Device<Kokkos::CudaSpace::execution_space, Kokkos::CudaSpace::memory_space>, Kokkos::MemoryTraits<0U>>, RT=Kokkos::complex<float> *****, RP=<Kokkos::LayoutRight, execution_space>, Arg0=std::pair<int, std::size_t>, Args=<std::pair<int, std::size_t>, std::pair<int, std::size_t>, std::pair<int, std::size_t>, std::pair<int, std::size_t>>]" at line 106 of /work/04/jh230064a/i18048/tests/develop/kokkos-fft/common/src/KokkosFFT_padding.hpp
            instantiation of "void KokkosFFT::Impl::crop_or_pad_impl(const ExecutionSpace &, const InViewType &, const OutViewType &, std::index_sequence<Is...>) [with ExecutionSpace=execution_space, InViewType=Kokkos::View<Kokkos::complex<float> *****, Kokkos::LayoutRight, execution_space>, OutViewType=Kokkos::View<Kokkos::complex<float> *****, Kokkos::LayoutRight, execution_space>, Is=<0UL, 1UL, 2UL, 3UL, 4UL>]" at line 122 of /work/04/jh230064a/i18048/tests/develop/kokkos-fft/common/src/KokkosFFT_padding.hpp
            instantiation of "void KokkosFFT::Impl::crop_or_pad(const ExecutionSpace &, const InViewType &, const OutViewType &) [with ExecutionSpace=execution_space, InViewType=Kokkos::View<Kokkos::complex<float> *****, Kokkos::LayoutRight, execution_space>, OutViewType=Kokkos::View<Kokkos::complex<float> *****, Kokkos::LayoutRight, execution_space>]" at line 935 of /work/04/jh230064a/i18048/tests/develop/kokkos-fft/fft/unit_test/Test_Transform.cpp
            instantiation of "void test_fft1_1dfft_5dview<T,LayoutType>(T) [with T=float, LayoutType=Kokkos::LayoutRight]" at line 1232 of /work/04/jh230064a/i18048/tests/develop/kokkos-fft/fft/unit_test/Test_Transform.cpp
            [ 7 instantiation contexts not shown ]
            instantiation of class "testing::internal::TestFactoryImpl<TestClass> [with TestClass=FFT1D_FFT_batched_5DView_Test<std::pair<float, Kokkos::LayoutRight>>]" at line 723 of /work/04/jh230064a/i18048/tests/develop/kokkos-fft/tpls/googletest/googletest/include/gtest/internal/gtest-internal.h
            implicit generation of "testing::internal::TestFactoryImpl<TestClass>::TestFactoryImpl() [with TestClass=FFT1D_FFT_batched_5DView_Test<std::pair<float, Kokkos::LayoutRight>>]" at line 723 of /work/04/jh230064a/i18048/tests/develop/kokkos-fft/tpls/googletest/googletest/include/gtest/internal/gtest-internal.h
            instantiation of class "testing::internal::TestFactoryImpl<TestClass> [with TestClass=FFT1D_FFT_batched_5DView_Test<std::pair<float, Kokkos::LayoutRight>>]" at line 723 of /work/04/jh230064a/i18048/tests/develop/kokkos-fft/tpls/googletest/googletest/include/gtest/internal/gtest-internal.h
            instantiation of "__nv_bool testing::internal::TypeParameterizedTest<Fixture, TestSel, Types>::Register(const char *, testing::internal::CodeLocation, const char *, const char *, int, const std::vector<std::string, std::allocator<std::string>> &) [with Fixture=FFT1D, TestSel=testing::internal::TemplateSel<FFT1D_FFT_batched_5DView_Test>, Types=testing::internal::Types<std::pair<float, Kokkos::LayoutRight>, std::pair<double, Kokkos::LayoutLeft>, std::pair<double, Kokkos::LayoutRight>>]" at line 726 of /work/04/jh230064a/i18048/tests/develop/kokkos-fft/tpls/googletest/googletest/include/gtest/internal/gtest-internal.h
            instantiation of "__nv_bool testing::internal::TypeParameterizedTest<Fixture, TestSel, Types>::Register(const char *, testing::internal::CodeLocation, const char *, const char *, int, const std::vector<std::string, std::allocator<std::string>> &) [with Fixture=FFT1D, TestSel=testing::internal::TemplateSel<FFT1D_FFT_batched_5DView_Test>, Types=gtest_type_params_FFT1D_]" at line 1227 of /work/04/jh230064a/i18048/tests/develop/kokkos-fft/fft/unit_test/Test_Transform.cpp

/work/04/jh230064a/i18048/tests/develop/kokkos-fft/tpls/kokkos/core/src/impl/Kokkos_ViewMapping.hpp(3352): error: no instance of constructor "Kokkos::Impl::ViewOffset<Dimension, Kokkos::LayoutRight, std::enable_if_t<<expression>, void>>::ViewOffset [with Dimension=Kokkos::Impl::ViewDimension<0UL, 0UL, 0UL, 0UL, 0UL>]" matches the argument list
            argument types are: (const Kokkos::Impl::ViewOffset<Kokkos::Impl::ViewDimension<0UL, 0UL, 0UL, 0UL, 0UL>, Kokkos::LayoutRight, void>, const Kokkos::Impl::SubviewExtents<5U, 0U>)
      dst.m_impl_offset = dst_offset_type(src.m_impl_offset, extents);
                          ^
          detected during:
            instantiation of "void Kokkos::Impl::ViewMapping<std::enable_if_t<<expression>, void>, SrcTraits, Args...>::assign(Kokkos::Impl::ViewMapping<DstTraits, void> &, const Kokkos::Impl::ViewMapping<SrcTraits, void> &, Args...) [with SrcTraits=Kokkos::ViewTraits<float *****, Kokkos::LayoutRight, execution_space>, Args=<std::pair<int, std::size_t>, std::pair<int, std::size_t>, std::pair<int, std::size_t>, std::pair<int, std::size_t>, std::pair<int, std::size_t>>, DstTraits=Kokkos::ViewTraits<float *****, Kokkos::LayoutRight, Kokkos::Device<Kokkos::CudaSpace::execution_space, Kokkos::CudaSpace::memory_space>, Kokkos::MemoryTraits<0U>>]" at line 1372 of /work/04/jh230064a/i18048/tests/develop/kokkos-fft/tpls/kokkos/core/src/Kokkos_View.hpp
            instantiation of "Kokkos::View<DataType, Properties...>::View(const Kokkos::View<RT, RP...> &, Arg0, Args...) [with DataType=float *****, Properties=<Kokkos::LayoutRight, Kokkos::Device<Kokkos::CudaSpace::execution_space, Kokkos::CudaSpace::memory_space>, Kokkos::MemoryTraits<0U>>, RT=float *****, RP=<Kokkos::LayoutRight, execution_space>, Arg0=std::pair<int, std::size_t>, Args=<std::pair<int, std::size_t>, std::pair<int, std::size_t>, std::pair<int, std::size_t>, std::pair<int, std::size_t>>]" at line 106 of /work/04/jh230064a/i18048/tests/develop/kokkos-fft/common/src/KokkosFFT_padding.hpp
            instantiation of "void KokkosFFT::Impl::crop_or_pad_impl(const ExecutionSpace &, const InViewType &, const OutViewType &, std::index_sequence<Is...>) [with ExecutionSpace=execution_space, InViewType=Kokkos::View<float *****, Kokkos::LayoutRight, execution_space>, OutViewType=Kokkos::View<float *****, Kokkos::LayoutRight, execution_space>, Is=<0UL, 1UL, 2UL, 3UL, 4UL>]" at line 122 of /work/04/jh230064a/i18048/tests/develop/kokkos-fft/common/src/KokkosFFT_padding.hpp
            instantiation of "void KokkosFFT::Impl::crop_or_pad(const ExecutionSpace &, const InViewType &, const OutViewType &) [with ExecutionSpace=execution_space, InViewType=Kokkos::View<float *****, Kokkos::LayoutRight, execution_space>, OutViewType=Kokkos::View<float *****, Kokkos::LayoutRight, execution_space>]" at line 936 of /work/04/jh230064a/i18048/tests/develop/kokkos-fft/fft/unit_test/Test_Transform.cpp
            instantiation of "void test_fft1_1dfft_5dview<T,LayoutType>(T) [with T=float, LayoutType=Kokkos::LayoutRight]" at line 1232 of /work/04/jh230064a/i18048/tests/develop/kokkos-fft/fft/unit_test/Test_Transform.cpp
            [ 7 instantiation contexts not shown ]
            instantiation of class "testing::internal::TestFactoryImpl<TestClass> [with TestClass=FFT1D_FFT_batched_5DView_Test<std::pair<float, Kokkos::LayoutRight>>]" at line 723 of /work/04/jh230064a/i18048/tests/develop/kokkos-fft/tpls/googletest/googletest/include/gtest/internal/gtest-internal.h
            implicit generation of "testing::internal::TestFactoryImpl<TestClass>::TestFactoryImpl() [with TestClass=FFT1D_FFT_batched_5DView_Test<std::pair<float, Kokkos::LayoutRight>>]" at line 723 of /work/04/jh230064a/i18048/tests/develop/kokkos-fft/tpls/googletest/googletest/include/gtest/internal/gtest-internal.h
            instantiation of class "testing::internal::TestFactoryImpl<TestClass> [with TestClass=FFT1D_FFT_batched_5DView_Test<std::pair<float, Kokkos::LayoutRight>>]" at line 723 of /work/04/jh230064a/i18048/tests/develop/kokkos-fft/tpls/googletest/googletest/include/gtest/internal/gtest-internal.h
            instantiation of "__nv_bool testing::internal::TypeParameterizedTest<Fixture, TestSel, Types>::Register(const char *, testing::internal::CodeLocation, const char *, const char *, int, const std::vector<std::string, std::allocator<std::string>> &) [with Fixture=FFT1D, TestSel=testing::internal::TemplateSel<FFT1D_FFT_batched_5DView_Test>, Types=testing::internal::Types<std::pair<float, Kokkos::LayoutRight>, std::pair<double, Kokkos::LayoutLeft>, std::pair<double, Kokkos::LayoutRight>>]" at line 726 of /work/04/jh230064a/i18048/tests/develop/kokkos-fft/tpls/googletest/googletest/include/gtest/internal/gtest-internal.h
            instantiation of "__nv_bool testing::internal::TypeParameterizedTest<Fixture, TestSel, Types>::Register(const char *, testing::internal::CodeLocation, const char *, const char *, int, const std::vector<std::string, std::allocator<std::string>> &) [with Fixture=FFT1D, TestSel=testing::internal::TemplateSel<FFT1D_FFT_batched_5DView_Test>, Types=gtest_type_params_FFT1D_]" at line 1227 of /work/04/jh230064a/i18048/tests/develop/kokkos-fft/fft/unit_test/Test_Transform.cpp
```

**Comment by yasahi-hpc** [comment]
This works

```C++
template <typename ExecutionSpace, typename InViewType, typename OutViewType,
          std::size_t... Is>
void crop_or_pad_impl(const ExecutionSpace& exec_space, const InViewType& in,
                      const OutViewType& out, std::index_sequence<Is...>) {
  constexpr std::size_t rank = InViewType::rank();
  using extents_type = std::array<std::size_t, rank>;

  extents_type extents;
  for (std::size_t i = 0; i < rank; i++) {
    extents.at(i) = std::min(in.extent(i), out.extent(i));
  }

  auto sub_in  = Kokkos::subview(in, std::make_pair(std::size_t(0), std::get<Is>(extents))...);
  auto sub_out = Kokkos::subview(out, std::make_pair(std::size_t(0), std::get<Is>(extents))...);
  Kokkos::deep_copy(exec_space, sub_out, sub_in);
}

template <typename ExecutionSpace, typename InViewType, typename OutViewType>
void crop_or_pad(const ExecutionSpace& exec_space, const InViewType& in,
                 const OutViewType& out) {
  static_assert(
      KokkosFFT::Impl::are_operatable_views_v<ExecutionSpace, InViewType,
                                              OutViewType>,
      "crop_or_pad: InViewType and OutViewType must have the same base "
      "floating point "
      "type (float/double), the same layout (LayoutLeft/LayoutRight), and the "
      "same rank. ExecutionSpace must be accessible to the data in InViewType "
      "and OutViewType.");
  crop_or_pad_impl(exec_space, in, out, std::make_index_sequence<InViewType::rank()>{});
}
```

**Comment by tpadioleau** [comment]
Seems good

**Review by tpadioleau — DISMISSED** [review]
One remaining question but looks globally good to me

File: common/src/KokkosFFT_utils.hpp (hunk 2 above)
Comment (yasahi-hpc) [reply common/src/KokkosFFT_utils.hpp:212]: Just in case, we want to use `std::vector`? 
Or do you think it is better not to allow that

Comment (tpadioleau) [reply common/src/KokkosFFT_utils.hpp:212]: I would generalize when needed

**Review by tpadioleau — APPROVED** [review]
LGTM

File: common/src/KokkosFFT_utils.hpp (hunk 2 above)
Comment (yasahi-hpc) [reply common/src/KokkosFFT_utils.hpp:212]: Sure, I will keep `std::array<IntType, N>` for the moment then.

**Comment by yasahi-hpc** [comment]
@tpadioleau
Thank you for the review
Due to the GPU CI issue, we cannot merge this for the moment.

**Comment by tpadioleau** [comment]
> @tpadioleau Thank you for the review Due to the GPU CI issue, we cannot merge this for the moment.

In "Set up job" of the failed GPU CI I see
> Warning: You are running out of disk space. The runner will stop working when the machine runs out of disk space. Free space left: 15 MB

**Comment by yasahi-hpc** [comment]
> > @tpadioleau Thank you for the review Due to the GPU CI issue, we cannot merge this for the moment.
> 
> In "Set up job" of the failed GPU CI I see
> 
> > Warning: You are running out of disk space. The runner will stop working when the machine runs out of disk space. Free space left: 15 MB

Oh, the disk space on Ruche? Then, we cannot do anything.
Maybe ignore the failure for the moment

**Comment by yasahi-hpc** [comment]
> Maybe let's ask @pzehner before merging

Sure
